lib/plugins/aws/package/compile/events/s3/index.js

```diff
@@ -74,21 +101,41 @@ class AwsCompileS3Events {
                 filter = { Filter: { S3Key: { Rules: rules } } };
               }
             } else if (typeof event.s3 === 'string') {
-              bucketName = event.s3;
+              bucketRef = event.s3;
             } else {
               const errorMessage = [
                 `S3 event of function ${functionName} is not an object nor a string.`,
-                ' The correct syntax is: s3: bucketName OR an object with "bucket" property.',
-                ' Please check the docs for more info.',
-              ].join('');
+                this.errorMessageBucketSyntax,
+                'Please check the docs for more info.',
+              ].join(' ');
               throw new this.serverless.classes.Error(errorMessage);
             }
 
             const lambdaLogicalId = this.provider.naming.getLambdaLogicalId(functionName);
+            let bucketName;
+            if (provderS3[bucketRef]) {
+              bucketName =
+                provderS3[bucketRef].name ||
+                this.provider.naming
+                  .normalizeBucketName(bucketRef)
+                  .replace(/\B([A-Z])/, '-$1')
+                  .toLowerCase();
```

Comment:
The reason why this is used is so we can ensure bucket names are valid when we generate them. This is only used when we use an empty object in `provider.s3.myBucket`. If i dont add this, the resulting bucking name would be `myBucket` which is an invalid name. This like would instead create a bucket with the name `my-bucket`

Reply:
> This is only used when we use an empty object in `provider.s3.myBucket`

Isn't it used when there's no `name` provided to  provider.s3.myBucket (so also when there are some other properties)?

Technically I think it'll be good to maintain similar convention as with function names (although function names allow capital letters, and framework also prefixes them with service and stage name).

Do you think it's an issue for a developer to ensure that used ref name is constructed only from characters that are allowed in bucket name?

Reply:
> Isn't it used when there's no `name` provided to provider.s3.myBucket (so also when there are some other properties)?

Yes.

> Technically I think it'll be good to maintain similar convention as with function names (although function names allow capital letters, and framework also prefixes them with service and stage name).

Should we add the same logic here (stage + service name)? I think we must at least lower case them since CF will error otherwise by just following yaml key naming convention (camel case).

> Do you think it's an issue for a developer to ensure that used ref name is constructed only from characters that are allowed in bucket name?

Yes.

Reply:
> Should we add the same logic here (stage + service name)? 

I wouldn't do it, as we don't do it with `bucketName` as put at `s3` event. Technically framework is already inconsistent, and trying to fix that will make a breaking change

>  I think we must at least lower case them since CF will error otherwise by just following yaml key naming convention (camel case).

The problem I see, is that if we mangle that name, then it becomes confusing. User cannot be sure what the final name of created bucket will be and how to find it.

AFAIK we don't do such mangling in other cases (aside of prefixing with stage and service name), so I'm not sure about introducing here.
I think we do such normalization in case of CF resource naming, but that's different thing (those ids are used only in CF template context)

@pmuens what do you think?

Reply:
@medikoo what's your suggestion then? Lowercase it only? :)

Reply:
I would pass as is 1:1, and let AWS template validation fail if invalid name is used.

Still let's wait for feedback from @pmuens

Reply:
> I would pass as is 1:1, and let AWS template validation fail if invalid name is used.
> 
> Still let's wait for feedback from @pmuens

I think you're missing the point here. The deploy will probably rarely succeed since keys in `provider.s3` are very likely to be camel-cased. This modification does only apply when no name or bucketName property exists in the `provider.s3.myBucket` object.

**The following will succeed:**
```yml
provider:
  s3:
     myBucket:
       name: my-awesome-custom-bucket
```

**This will not (with your suggestion)**
```yml
provider:
  s3:
     myBucket:
       objectLockEnabled: true
```
_(☝️ the bucket name will be "myBucket" without the modifications I have made, which is invalid since it contains uppercase chars)_

Reply:
@richarddd Thanks for explanations, Still I have full understanding of that, and my feeling is that it's better if user is forced to configure bucket names 1:1 as follows:

```yaml
provider:
  s3:
    myBucket:
      name: my-awesome-custom-bucket
      objectLockEnabled: true
```

Or

```yaml
provider:
  s3:
    my-awesome-custom-bucket:
      objectLockEnabled: true
```

And have full control and understanding of what final bucket name will be (no magical bucket name correction)

Being at stake of normalization algorithm (which while will allow him to put any string as bucket reference), will in many cases create a bucket with name he didn't put (I'm worried in many cases it'll turn confusing and surprising).

--- 

Note that Framework currently is not doing that. If someone will put:

```yaml
functions:
  function:
    handler: index.handler
    events:
      - s3: myBucket
```

Deployment will crash with ` An error occurred: S3BucketTestFoo - Bucket name should not contain uppercase characters.`

And I think it's better than implying normalization.

If we agree on what you're proposing, then just adding following to above config

```yaml
provider:
    s3:
      myBucket: {}
```

Will change that behavior, and create S3 with `mybucket` name. That doesn't feel right to me.

--- 

I understand that your argument is to allow users to use camelCase (natural) notation in YAML for bucket configuration, but I think  predictability of the outcome is more valuable here, then maintaining config file property names convention.

Reply:
Other potential issue with normalization implied is that different refs may translate to same bucketName (e.g. `myBucket`, `mybucket`, `MY_BUCKET` etc will resolve to `mybucket`) causing collisions, which are probably not handled well.

Reply:
@medikoo you have a valid point there! I came to think of swagger v3 spec where the routes/paths are indeed a 1:1 which feels quite natural. I'll update the PR and remove the normalization :)

Reply:
btw, do you think we should remove this check also? I feel like maybe this would be great to catch in an early stage:  
https://github.com/serverless/serverless/pull/7156/files#diff-a5a4c36cd72d14b90ae83a1ed4bb1fdbR123

Reply:
Good question, I agree it'll be convenient to abort early in invalid name case. I would  just move it out of condition, so it's validated in all configuration cases

Reply:
Thanks for the feedback. Fixed in the latest push!